crates/driver/src/settlement_proposal.rs

```diff
@@ -0,0 +1,356 @@
+#[derive(Debug, Default, Clone)]
+pub struct SettlementProposal {
+    pub clearing_prices: HashMap<H160, U256>,
+    pub trades: Vec<TradedOrder>,
+    pub execution_plan: Vec<Arc<dyn InteractionProposal>>,
+}
+
+impl SettlementProposal {
+    /// Calculates the surplus generated by this `SettlementProposal` denominated in the native
+    /// token.
+    pub fn surplus(&self, external_prices: &ExternalPrices) -> Result<BigRational> {
+        self.trades.iter().fold(Ok(num::zero()), |acc, trade| {
+            let normalized_surplus = trade_surplus_in_native_token(
+                &trade.order,
+                trade.executed_amount,
+                external_prices,
+                &self.clearing_prices,
+            )
+            .context("could not compute surplus for trade")?;
+            Ok(acc? + normalized_surplus)
+        })
+    }
+
+    /// Turns the proposal into a `SettlementEncoder` which contains finalized call data for all
+    /// the interactions.
+    pub async fn into_encoder(self) -> Result<SettlementEncoder> {
+        let mut encoder = SettlementEncoder::new(self.clearing_prices);
+
+        for trade in self.trades {
+            let remaining_amounts = trade.order.remaining_amounts()?;
+
+            if trade.order.metadata.is_liquidity_order {
+                encoder.add_liquidity_order_trade(
+                    trade.order,
+                    trade.executed_amount,
+                    remaining_amounts.fee_amount,
+                )?;
+            } else {
+                encoder.add_trade(
+                    trade.order,
+                    trade.executed_amount,
+                    remaining_amounts.fee_amount,
+                )?;
+            }
+        }
+
+        let futures = self
+            .execution_plan
+            .iter()
+            .map(|interaction| interaction.finalize());
+        let interactions = futures::future::try_join_all(futures).await?;
+        interactions
+            .into_iter()
+            .flat_map(|i| i.encode())
+            .for_each(|i| encoder.append_to_execution_plan(i));
+
+        Ok(encoder)
+    }
+
+    pub async fn into_settlement(self) -> Result<Settlement> {
+        Ok(Settlement {
+            encoder: self.into_encoder().await?,
+        })
+    }
+
+    /// Computes the `SettlementSummary` if following checks are successful:
+    ///   - individual trades don't violate required properties
+    ///   - enough token balances before each on-chain interaction
+    ///   - enough token balances to pay out orders at the end
+    ///   - token conservation (TODO)
+    pub fn into_settlement_summary(
+        &self,
+        gas_price: f64,
+        external_prices: &ExternalPrices,
+    ) -> Result<SettlementSummary> {
+        let mut balances = HashMap::<H160, U256>::default();
+        let mut gas_used = U256::zero();
+
+        let mut trade_executions = Vec::with_capacity(self.trades.len());
```

Comment:
You should be able to collect an `Iterator<Item = Result<T, E>>` into a `Result<Vec<T>, E>`. No issues with the `for` loop if you feel like its cleaner.